### PR TITLE
Rust: Add Missing Elements query

### PR DESCRIPTION
**Pull request by geoffw0**
Add Missing Elements diagnostic query (`rust/diagnostics/missing-elements`) that lists `Unimplemented`, `MissingExpr` and `MissingPat` elements.  At the moment there's rather a lot, even on the simple test - probably mostly coming from the standard libraries.  Future extractor changes should greatly reduce this.

### CODE REVIEW

**Comment by geoffw0** [comment]
Hmm, the locations need cleaning up here...

File: rust/ql/src/queries/diagnostics/MissingElements.ql [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import rust
+
+/**
+ * Gets `l.toString()`, but with any locations outside of the source location prefix cleaned up.
+ */
+bindingset[l]
+string cleanLocationString(Location l) {
+  if exists(l.getFile().getRelativePath()) or l instanceof EmptyLocation
+  then result = l.toString()
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply rust/ql/src/queries/diagnostics/MissingElements.ql:15]: It's fine, it's only used for output and de-duplication of output.

File: rust/ql/src/queries/diagnostics/MissingElements.ql [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+  else l.getFile().getParentContainer().getAbsolutePath() + result = l.toString() // remove the directory from the string
```
[the remark this answers is not in the document]
Comment (geoffw0) [reply rust/ql/src/queries/diagnostics/MissingElements.ql:16]: See previous.

**Comment by geoffw0** [comment]
OK, we get different results on CI to my run locally.  I think this will be a continuous source of problems unless we either (1) restrict results to inside the source location or (2) remove or constrain the test.  (1) seems reasonable but it does mean we won't have visibility on problems extracting the standard library.

I'll have another think about it on Monday.  Opinions welcome.

**Comment by aibaars** [comment]
> OK, we get different results on CI to my run locally. I think this will be a continuous source of problems unless we either (1) restrict results to inside the source location or (2) remove or constrain the test. (1) seems reasonable but it does mean we won't have visibility on problems extracting the standard library.
> 
> I'll have another think about it on Monday. Opinions welcome.

My other PR no longer extracts files in the standard library or dependencies so that problem should be solved. I added a flag to enable the extraction of libraries but by default we no longer do. Extracting the standard library is too brittle because results depend on whichever version of rust is installed and most likely the platform on which it runs also causes differences.

**Comment by aibaars** [comment]
@geoffw0 @redsun82 This PR has a compilation error because `class Unimplemented(Element)` does not have a `getLocation` method. We could either change the schema to have `class Unimplemented(Locatable)` or add a predicate like `Location getLocation(Unimplemented node) { result = node.(Locatable).getLocation() or result instanceof UnknownLocation and not node instanceof Locatable}|` to this PR. Not sure what is best, I'm fine with either approach.

**Comment by geoffw0** [comment]
> My other PR no longer extracts files in the standard library or dependencies so that problem should be solved. I added a flag to enable the extraction of libraries but by default we no longer do. Extracting the standard library is too brittle because results depend on whichever version of rust is installed and most likely the platform on which it runs also causes differences.

Does that mean we'll have difficulty assigning targets for calls to standard library functions?  Or are we extracting *some* stuff in the standard libraries still?

File: rust/ql/src/queries/diagnostics/MissingElements.ql [hunk 3]
```diff
@@ -39,23 +29,33 @@ string multipleString(int i) {
 query predicate listUnimplemented(string location, string msg) {
   // something that is not extracted yet
   exists(int c |
-    c = strictcount(Unimplemented n | cleanLocationString(getUnimplementedLocation(n)) = location) and
+    c =
+      strictcount(Unimplemented n |
+        exists(getUnimplementedLocation(n).getFile().getRelativePath()) and
```
Comment (aibaars) [inline rust/ql/src/queries/diagnostics/MissingElements.ql:34]: This excludes elements from user code for which the extractor did not know a location.

Comment (aibaars) [reply rust/ql/src/queries/diagnostics/MissingElements.ql:34]: The MissingElement query is a bit special. It's purpose is to measure the quality of the database/extractor. If we throw out all MissingElements without location then we might think a database is in perfect health while in reality it is not.

Comment (geoffw0) [reply rust/ql/src/queries/diagnostics/MissingElements.ql:34]: ... and we're committed to *not* extracting non-user code now as I understand it (still now 100% sure how that works).  I'll try including them again.

Comment (geoffw0) [reply rust/ql/src/queries/diagnostics/MissingElements.ql]: I've just reverted the restriction to user code in this PR as well.  I'm fairly sure I was seeing inconsistencies with it though, we will see...

Comment (geoffw0) [reply rust/ql/src/queries/diagnostics/MissingElements.ql]: Yep.
```
-| @0:0:0:0 | Not yet implemented (x20). |
+| @0:0:0:0 | Not yet implemented (x14). |
```
This is the difference between my local results and CI.  Is this something we can fix in the extractor quickly, or should I perhaps change the PR to align with CI (in which case we can expect local test failures until the issue is fixed)?

Comment (aibaars) [reply rust/ql/src/queries/diagnostics/MissingElements.ql]: I think I found the source of the trouble. Could you try removing the use of the `println!` macro? I suspect that macro gets expanded differently depending on the rust version/target platform, or whatever other difference in environment .

Comment (geoffw0) [reply rust/ql/src/queries/diagnostics/MissingElements.ql]: That reduces me to "Not yet implemented (x17)" locally.

Comment (aibaars) [reply rust/ql/src/queries/diagnostics/MissingElements.ql]: Did you comment out the`println!` in `my_macro.rs` as well as `main.rs`?

Comment (geoffw0) [reply rust/ql/src/queries/diagnostics/MissingElements.ql]: ... oops ...

Yep, that gets me down to 14, which is presumably in line with CI now.  I've pushed the fix.  There *is* still an issue in the extractor presumably, but this PR should be mergeable now I think.

Comment (aibaars) [reply rust/ql/src/queries/diagnostics/MissingElements.ql]: Thanks, yes there must be some issue, or perhaps the standard library for rust on the CI runners has a different implementation for `println!`.

Comment (redsun82) [reply rust/ql/src/queries/diagnostics/MissingElements.ql]: hmm, that is unfortunate. But maybe we just need to do a https://github.com/actions-rust-lang/setup-rust-toolchain step for running the QL tests?

**Review by redsun82 — APPROVED** [review]
I'm ok with this going in as it is, it will be valuable for us!

Can you open an issue to track the problem with `println!` in tests leading to differences in extraction?

**Comment by geoffw0** [comment]
> Can you open an issue to track the problem with println! in tests leading to differences in extraction?

Done (provisionally assigned to you).

> the name of the query might be better using unextracted rather than missing for consistency.

Will do...

**Comment by geoffw0** [comment]
> the name of the query might be better using unextracted rather than missing for consistency.

Done.

**Pull request merged**